File: 04-Monitoriai/counter/src/edu/ktu/ryselis/SimpleBoundedCounter.java
package edu.ktu.ryselis;

class SimpleBoundedCounter {
    private final long MIN = 0;
    private final long MAX = 50;
    private long count = MIN;

    synchronized void increase() throws InterruptedException {
        awaitUnderMax();
        setCount(count + 1);
    }

    synchronized void decrease() throws InterruptedException {
        awaitOverMin();
        setCount(count - 1);
    }

    private void setCount(long newValue){
        count = newValue;
        notifyAll();
    }

    private void awaitUnderMax() throws InterruptedException {
        while (count == MAX) {
            wait();
        }
    }

    private void awaitOverMin() throws InterruptedException {
        while (count == MIN) {
            wait();
        }
    }

    long getCount(){
        return count;
    }
}

File: 05-Monitoriai-cpp-cs-openmp/counter_cs/ConsoleApplication1/ConsoleApplication1/SimpleBoundedCounter.cs
using System.Threading;

namespace ConsoleApplication1
{
    public class SimpleBoundedCounter
    {
        private const long Min = 0;
        private const long Max = 50;
        private long _count;

        public long Count
        {
            get => _count;
            private set
            {
                _count = value;
                Monitor.PulseAll(_locker);
            }
        }

        private readonly object _locker;

        public SimpleBoundedCounter()
        {
            _count = Min;
            _locker = new object();
        }

        public void Increase()
        {
            lock (_locker)
            {
                while (Count == Max)
                {
                    Monitor.Wait(_locker);
                }
                Count = Count + 1;
            }
        }

        public void Decrease()
        {
            lock (_locker)
            {
                while (Count == 0)
                {
                    Monitor.Wait(_locker);
                }

                Count = Count - 1;
            }
        }
    }
}
